include/swift/AST/LayoutConstraint.h

```diff
@@ -0,0 +1,107 @@
+//===-- LayoutConstraint.h - Layout constraints types and APIs --*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines types and APIs for layout constraints.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_LAYOUT_CONSTRAINT_H
+#define SWIFT_LAYOUT_CONSTRAINT_H
+
+#include "llvm/ADT/StringRef.h"
+
+namespace swift {
+
+class Type;
+class TypeRepr;
+
+/// Describes a layout constraint information.
+enum class LayoutConstraintKind {
+  // It is not a known layout constraint.
+  UnknownLayout,
+  Trivial8 = 1,
```

Comment:
@slavapestov Btw, @rjmccall had the idea to allow for layout constraints like `_TrivialOfSizeAtMostN`, because it can be very handy. I'll see what I can do. But it is an additive feature I could add later.